### ISSUE TITLE
Rule Proposal: lines-around-directive

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#6069]
From [requirePaddingNewLinesAfterUseStrict] and [disallowPaddingNewLinesAfterUseStrict].

This rule will require/disallow blank lines before/after a chunk of directive prologues.

```js
{
    "lines-around-directive": ["error", "always" or "never"],
    // or
    "lines-around-directive": ["error", {
        "after": "always" or "never",
        "before": "always" or "never"
    }]
}
```

**Valid:**

```js
/*eslint lines-around-directive: ["error", "always"]*/

// header comment?

"use strict";

foo();

function foo() {
    "use strict";  // ignore at the head of block. this is warned by padded-blocks.
    "use asm";     // ignore between directives.

    foo();
}
```

```js
/*eslint lines-around-directive: ["error", "never"]*/

// header comment?
"use strict";
foo();

function foo() {

    "use strict";  // ignore at the head of block. this is warned by padded-blocks.
    "use asm";     // ignore between directives.
    foo();
}
```

**Invalid:**

```js
/*eslint lines-around-directive: ["error", "always"]*/

// header comment?
"use strict";
foo();

function foo() {
    "use strict";
    foo();
}
```

```js
/*eslint lines-around-directive: ["error", "never"]*/

// header comment?

"use strict";

foo();

function foo() {
    "use strict";

    foo();
}
```

[requirePaddingNewLinesAfterUseStrict]: http://jscs.info/rule/requirePaddingNewLinesAfterUseStrict
[disallowPaddingNewLinesAfterUseStrict]: http://jscs.info/rule/disallowPaddingNewLinesAfterUseStrict

**Comment by mysticatea** [comment]
I'm idiot! `before` is nothing.

**Comment by mysticatea** [comment]
Can we check lines between Shebang and directive prologues?

**Comment by ilyavolodin** [comment]
@mysticatea We should be able to, since shebang is commented out before parsing, we just have to check comment value.

**Comment by nzakas** [comment]
Do we need to check padding before? That would almost always be covered by `padded-blocks`, wouldn't it?

**Comment by mysticatea** [comment]
I had imaged the previous padding for shebang and header comment.

```js
#!/usr/bin/env node

"use strict"
```
 
```js
/**
 * header comment
 */

"use strict"
```

**Comment by nzakas** [comment]
I see. So then this rule's `before` would not apply in blocks?

**Comment by mysticatea** [comment]
```js
/*eslint lines-around-directives*/

function a() {
    "use strict";
}

function a() {

    "use strict";
}

function a() {
    /*
     * comment
     */

    "use strict";
}
```

**Comment by mysticatea** [comment]
Sorry, I click a wrong button...

I'm thinking `before` would apply if there is a comment at the top of blocks.

**Comment by nzakas** [comment]
Doesn't `padded-blocks` already cover that case, too?

**Comment by mysticatea** [comment]
`padded-blocks` checks lines before the top comment in the block.

```js
function foo() {
    <!-- by padded-blocks -->
    /*COMMENT*/
    <!-- by lines-around-directives -->
    "use strict";
    <!-- by lines-around-directives -->
    doSomething()
    <!-- by padded-blocks -->
}

**Comment by nzakas** [comment]
Ah, I see. Got it. 

👍

**Comment by kaicataldo** [comment]
Anything else left to discuss here? I'm looking for a JSCS compat issue to work on this weekend and this seems like a good one :)

**Comment by platinumazure** [comment]
:+1: from me.

**Comment by alberto** [comment]
👍

**Comment by mikesherov** [comment]
:+1:
